### ISSUE TITLE
.NET 5 RC1: Console.ReadKey inside of thread cannot read ENTER key (regression)

### ISSUE THREAD

**Issue by SommerEngineering** [dotnet/runtime#42333]
### Description
Assumptions: A console program reads key strokes from a user in a thread. However, it first checks whether a key has already been pressed.

In .NET 3.1 LTS everything works as it should. With the new .NET 5.0 RC1 there is at least one serious bug: The `ENTER` key is no longer recognized correctly. Instead of `ENTER` .NET 5 RC1 recognizes the `J` key.

It is very easy to reproduce this scenario. I have set up a public repository for it: https://github.com/SommerEngineering/BugEnterKey01

After the start, the program waits for key strokes. The console displays the integer value of the char and the recognized key. The program can be terminated with `ESC`.

By default the program runs with .NET 5.0 RC1. The bug is then immediately visible when the `ENTER` key is pressed several times. Interesting aspect: The bug is not deterministic. It happens that from time to time the key is recognized correctly.

Also interesting: The issue occurs only if a check is made beforehand to see if a key has been pressed [cf. line 20](https://github.com/SommerEngineering/BugEnterKey01/blob/e70bf38cd5b51ad4602c1d294d9d29176343dcf2/BugEnterKey01/Program.cs#L20).

Cross-checking with .NET 3.1 LTS is very easy. For this purpose, only one line in the project file needs to be commented out respectively activated again -- [cf. line 5 + 6](https://github.com/SommerEngineering/BugEnterKey01/blob/e70bf38cd5b51ad4602c1d294d9d29176343dcf2/BugEnterKey01/BugEnterKey01.csproj#L5-L6)

### Configuration
.NET 3.1.8 (SDK 3.1.402)
.NET 5.0 RC1 (5.0.100-rc.1.20452.10)
OS: Ubuntu Linux 20.04.1 LTS
CPU architecture: x64

**I investigated the issue so far only under Linux and cannot make a statement about Windows and macOS.**

### Regression?
Yes. As mentioned, it works fine with .NET 3.1.8.

### Other information

Possible workaround:
```
var key = Console.ReadKey();
var isEnter = (key.KeyChar == 10 && key.Key == ConsoleKey.J) || key.Key == ConsoleKey.Enter;
```

Is this might related to https://github.com/dotnet/runtime/issues/802?

**Comment by tmds** [comment]
This is probably related to changes in https://github.com/dotnet/runtime/pull/37491. cc @eiriktsarpalis

**Comment by tmds** [comment]
cc @danmosemsft

**Comment by tmds** [comment]
@SommerEngineering I took your repro app and removed the Thread.
```cs

using System;
using System.Threading;

namespace BugEnterKey01
{
    public static class Program
    {
        public static void Main(string[] args)
        {
            while (true)
            {
                if (Console.KeyAvailable)
                {
                    var key = Console.ReadKey(true);
                    Console.WriteLine($"Key info: char={(int)key.KeyChar}; recognized key={key.Key}");
                    
                    if(key.Key == ConsoleKey.Escape)
                        break;
                }
                
                Thread.Sleep(100);
            }
        }
    }
}
```
The issue reproduces for me:
```
$ bin/Debug/net5.0/console 
Key info: char=10; recognized key=J
Key info: char=10; recognized key=J
Key info: char=10; recognized key=J
Key info: char=10; recognized key=J
Key info: char=10; recognized key=J
Key info: char=10; recognized key=J
Key info: char=13; recognized key=Enter
Key info: char=13; recognized key=Enter
Key info: char=13; recognized key=Enter
```

**Comment by eiriktsarpalis** [comment]
I can reproduce it as well. Note that this is unrelated to `Thread.Sleep`, but most likely relates to the `Console.KeyAvailable` check. I removed that and the issue went away completely. 

I couldn't make the issue reproduce on WSL2.

**Comment by tmds** [comment]
> most likely relates to the Console.KeyAvailable check.

Yes, while checking `KeyAvailable`, the terminal is reconfigured which causes `\r` to be converted to `\n`. So `Enter` keys pressed during `KeyAvailable` check get returned as `Key info: char=10; recognized key=J`.

Probably the 0 argument should have been added to the front of the SystemNative_InitializeConsoleBeforeRead arguments instead of the back in https://github.com/dotnet/runtime/pull/37491. @eiriktsarpalis does that make sense?

https://github.com/dotnet/runtime/blob/873d963d37c82247b85dd510fc12167ee3077d39/src/libraries/Native/Unix/System.Native/pal_console.c#L379-L381

**Comment by tmds** [comment]
> Probably the 0 argument should have been added to the front of the SystemNative_InitializeConsoleBeforeRead arguments instead of the back in #37491. @eiriktsarpalis does that make sense?

I've made the change and it consistently reports Enter:
```
./artifacts/bin/testhost/net5.0-Linux-Debug-x64/shared/Microsoft.NETCore.App/6.0.0/corerun /tmp/console/bin/Debug/net5.0/console.dll
Key info: char=13; recognized key=Enter
Key info: char=13; recognized key=Enter
Key info: char=13; recognized key=Enter
Key info: char=13; recognized key=Enter
Key info: char=13; recognized key=Enter
Key info: char=13; recognized key=Enter
Key info: char=13; recognized key=Enter
Key info: char=13; recognized key=Enter
Key info: char=13; recognized key=Enter
Key info: char=13; recognized key=Enter
Key info: char=13; recognized key=Enter
Key info: char=13; recognized key=Enter
Key info: char=13; recognized key=Enter
Key info: char=13; recognized key=Enter
Key info: char=13; recognized key=Enter
Key info: char=13; recognized key=Enter
Key info: char=13; recognized key=Enter
Key info: char=13; recognized key=Enter
Key info: char=13; recognized key=Enter
Key info: char=13; recognized key=Enter

```

**Comment by eiriktsarpalis** [comment]
> does that make sense?

Yes, this is a very clear bug introduced [here](https://github.com/dotnet/runtime/commit/a62bfcbd6e03396e85232a73aeef10df9df5fa4d#diff-8885fe503b126dee072a443356a8a1baR382). It should have read 

```
SystemNative_InitializeConsoleBeforeRead(0, 1, 0);
```

**Comment by tmds** [comment]
I'm going to make a PR for it in a sec.

**Comment by eiriktsarpalis** [comment]
Thanks. @danmosemsft we should consider backporting this to 5.0.

**Comment by tmds** [comment]
@SommerEngineering thanks for the clear bug report!

**Comment by SommerEngineering** [comment]
Wow... thanks all of you for the fast fix and backport to 5.0 :slightly_smiling_face:

**Comment by eiriktsarpalis** [comment]
@SommerEngineering thank you for working out the minimal reproduction. Much appreciated 👍
